added cron tutorial

macolso opened the pull request:
Content must go through a pre-merge checklist.

Resolves https://github.com/fermyon/a3000-docs/issues/228 

Blocked on https://github.com/fermyon/developer/issues/1514

## Pre-Merge Content Checklist

This documentation has been checked to ensure that:

- [ ] The `title`, `template`, and `date` are all set
- [ ] Does this PR have a new menu item (anywhere in `templates/*.hbs` files) that points to a document `.md` that is set to publish in the future? If so please only publish the `.md` and `.hbs` changes in real-time (otherwise there will be a menu item pointing to a `.md` file that does not exist)
- [ ] File does not use CRLF, but uses plain LF (hint: use `cat -ve <filename> | grep $'\r' | wc -l` and expect 0 as a result)
- [ ] Has passed [`bart check`](https://developer.fermyon.com/bartholomew/quickstart)
- [ ] Has been manually tested by running in Spin/Bartholomew (hint: use `PREVIEW_MODE=1` and run `npm run styles` to update styling)
- [ ] Headings are using Title Case
- [ ] Code blocks have the programming language set to properly highlight syntax and the proper copy directive
- [ ] Have tested with `npm run test` and resolved all errors
- [ ] Relates to an existing (potentially outdated) blog article? If so please add URL in blog to point to this content.

itowlson commented:
FWIW I would not block this on updating the command reference but we would want to update the plugin install instructions so people get the 0.3 version.

itowlson commented on content/wasm-functions/using-cron-trigger.md:
````diff
@@ -32,16 +34,158 @@ Use Cases for Cron Jobs:
+<!-- @nocpy -->
+
+```console
+Hello from a cron component at 2025-03-31T21:45:25.562Z
````
Having these as separate blocks kind of disconnects the response from the curl command.  Maybe rephrase along the lines of:

> Test the app to make sure it's working correctly by running:
> (curl block)
> You should see a response along the lines of:
> (response block)

itowlson commented on content/wasm-functions/using-cron-trigger.md:
````diff
@@ -32,16 +34,158 @@ Use Cases for Cron Jobs:
+spin build --up
+```
+
+If it is working as expected, you should see a message returned along the lines of (but with an accurate datetime):
````
To me this reads like I should see the message from `spin up`, but you actually mean when I curl it.

itowlson commented on content/wasm-functions/using-cron-trigger.md:
```diff
@@ -32,16 +34,158 @@ Use Cases for Cron Jobs:
+
+## Deploy Spin Application to Fermyon Wasm Functions
+
+Before applying our cron trigger, we must have our Spin application running on the Fermyon Wasm Functions platform. Deploy the application with:
```
Reword cron trigger

itowlson commented on content/wasm-functions/using-cron-trigger.md:
```diff
@@ -32,16 +34,158 @@ Use Cases for Cron Jobs:
+
+# Creating a Cron Job for Your Spin Application
+
+Now we will use the `spin aka crons` command to invoke the Spin app's HTTP endpoint at our desired interval. The syntax for this request follows standard crontab format.
```
```suggestion
Now we will use the `spin aka crons` command to invoke the Spin app's HTTP endpoint on our desired schedule. The syntax for this request follows standard crontab format.
```

itowlson replied:
Link to the dialect of the crontab format that we support

itowlson commented on content/wasm-functions/using-cron-trigger.md:
```diff
@@ -32,16 +34,158 @@ Use Cases for Cron Jobs:
+
+Now we will use the `spin aka crons` command to invoke the Spin app's HTTP endpoint at our desired interval. The syntax for this request follows standard crontab format.
+
+> `spin aka cron` supports single digits in each cron expression field and intervals in every position except the day of the week (e.g., `* */12 * * *` for every 12 hours). The smallest supported cron configuration is every 5 minutes (`*/5 * * * *`). Other syntax elements like comma-separated lists and ranges are not supported. Please feel free to[leave feedback](https://fibsu0jcu2g.typeform.com/ ...
```
Is "single digits" correct?  You can't use double-digit numbers for e.g. day of month or hour of day?

itowlson commented on content/wasm-functions/using-cron-trigger.md:
```diff
@@ -32,16 +34,158 @@ Use Cases for Cron Jobs:
+
+> `spin aka cron` supports single digits in each cron expression field and intervals in every position except the day of the week (e.g., `* */12 * * *` for every 12 hours). The smallest supported cron configuration is every 5 minutes (`*/5 * * * *`). Other syntax elements like comma-separated lists and ranges are not supported. Please feel free to[leave feedback](https://fibsu0jcu2g.typeform.com/ ...
+
+Let's create a cron job that runs every hour. Since scheduling is based on UTC, you may need to [convert your local time to UTC](https://www.worldtimebuddy.com/?pl=1&lid=100&h=100&hf=1) to ensure correct execution.
```
I'm not sure why local/UTC would matter for an "every hour" schedule?  I'm _guessing_ it's that the schedule is "every hour _on_ the hour" and you're thinking about half-hour timezone offsets?  Like okay I convert my local time to UTC, it's 5:08am or whatever, now what do I do with that result?

It may be better to pull out the "times are in UTC" comment to go with the scheduling/crontab discussion perhaps?

itowlson commented on content/wasm-functions/using-cron-trigger.md:
````diff
@@ -32,16 +34,158 @@ Use Cases for Cron Jobs:
+<!-- @selectiveCpy -->
+
+```console
+spin aka crons create "0 * * * *" "/hello" "hello-cron-job"
````
Maybe explain what these fields are, since they are not named in the command.  I'm guessing the second is the route to be hit.  Is the third the name of the app to be hit?  Or the name of the job?  (In which case how does it know which app to hit?  Is this an "app in current directory" thing?)

If the third parameter is _not_ the name of the app, it would be good to choose a different string from the name of the app, so as to avoid ambiguity.

macolso commented on content/wasm-functions/using-cron-trigger.md:
```diff
@@ -7,18 +7,20 @@ enable_shortcodes = true
+- [Next Steps](#next-steps)
   
-This tutorial guides you through the process of scheduling HTTP requests in a Spin application running on Fermyon Wasm Functions using cron jobs. The `spin aka cron` command allows applications to receive scheduled HTTP triggers without relying on external services. In this tutorial, you'll learn how to define and manage cron jobs using the `spin aka cron` command.
+This tutorial guides you through the process of scheduling HTTP requests in a Spin application running on Fermyon Wasm Functions using cron jobs. The `spin aka crons` command allows applications to receive scheduled HTTP triggers without relying on external services. In this tutorial, you'll learn how to define and manage cron jobs using the `spin aka crons` command.
```
I also lean towards `cron` -- feels a bit more natural.

Also a good reminder I should flag that this feature is experimental in the tutorial

macolso commented on content/wasm-functions/using-cron-trigger.md:
```diff
@@ -32,16 +34,158 @@ Use Cases for Cron Jobs:
+# Create Spin application 
 
-Spin is a lightweight framework for building WebAssembly microservices and applications. It allows developers to define cron jobs that execute functions at scheduled intervals. To simplify development, Spin provides templates for various use cases, including scheduled tasks. 
+You must have a running Spin application on Fermyon Wasm Functions to apply the cron trigger to. We'll create a new Spin application for scheduling tasks using cron jobs using the `http-js` template:
```
You're right - it's not intended for scheduling tasks but the working is a bit misleading. I've reworded this sentence.

itowlson commented:
@macolso it occurred to me that one question with this is "how do I know it worked?"  What does the reader see if they follow the tutorial?  Can they see the HTTP responses in cron logs, for example?

(If not, what is the simplest example app that they can cron and still see an observable result?  A silly example might be: an app with two routes: `/cron-me` increments a value in KV storage, and `/count-me` responds with a message like "I've been cronned N times."  So the cron job could hit `/cron-me`, and the user could curl `/count-me` to see the value increment over time.  I realise this makes the tutorial more complicated, and distracts from the focus on the cron job though.)

macolso commented:
@itowlson I had a similar niggle with this tutorial. You can actually see the HTTP responses in the cron logs! I've updated the sample application to log cron events, inferring whether it's a cron event or a direct ping based on the presence of a query parameter. Let me know if this feels like a hacky solution, and I can remove that nuance.

itowlson commented on content/wasm-functions/using-cron-jobs.md:
````diff
@@ -0,0 +1,227 @@
+
+```console
+spin aka logs -a hello-cron-job
+2025-04-02 04:00:00 [hello-cron-job]  Hello, cron job triggered with msg: "fwf" at 2025-04-12T04:00:00Z
````
Consider linking this back to the `console.log` statement from from the HTTP application, and maybe put some more motivation around this, both here and when you write the application.  E.g. motivating the application a la "For a real cron job, you'd see its effects in the data it updates or the artifacts it produces. For this tutorial, though, we'll just emit a log, so that we can see the application has been called." and then down here we can say "This is the message we had the application log so that we could see it was being called" or something like that.

(If you keep the "different behaviour when they try it interactively" thing then it would be good to be _extra_ clear about this because the message will be unfamiliar.)

macolso replied:
Great suggestion! I added a disclaimer above the source code snippet that this is a simplified example and allude to what operations you may see in a real world example.

itowlson commented on content/wasm-functions/using-cron-jobs.md:
````diff
@@ -0,0 +1,227 @@
+Created cron which will next run at 2025-04-02 04:00:00 UTC
+```
+
+If we wait 5 minutes and check our application logs with the `spin aka logs` subcommand, we will see that our cron job invoked the application at the top of the hour as expected. 
````
"top of the hour" vs "every 5 minutes"

itowlson commented on content/wasm-functions/using-cron-jobs.md:
```diff
@@ -0,0 +1,227 @@
+
+* **Schedule:** How often the job runs, using standard cron syntax
+* **Path and query:** The HTTP path (and optional query parameters) to invoke
+* **Name:** The Spin applciaton to apply the cron job to
```
```suggestion
* **Name:** The Spin application to apply the cron job to
```

itowlson commented on content/wasm-functions/using-cron-jobs.md:
````diff
@@ -0,0 +1,227 @@
+<!-- @selectiveCpy -->
+
+```console
+spin aka crons create "*/5 * * * *" "/cron-me?msg=fwf" "hello-cron-job"
````
Oh man that's a lot of positional parameters and people are going to find it very easy to mess up the order.  I hope we give good error messages if they do.

itowlson commented on content/wasm-functions/using-cron-jobs.md:
```diff
@@ -0,0 +1,227 @@
+
+* **Schedule:** How often the job runs, using standard cron syntax
+* **Path and query:** The HTTP path (and optional query parameters) to invoke
+* **Name:** The Spin applciaton to apply the cron job to
```
The plugin docs say that the third parameter is the name of the _cron job_ not the name of the application.

(If that's true, we need to explain which application it's going to hit - I assume the one in the current directory.)

_(ETA: see Ryan's earlier comment https://github.com/fermyon/developer/pull/1515#discussion_r2024352152)_

macolso replied:
Good catch. I've updated accordingly.

itowlson commented on content/wasm-functions/using-cron-jobs.md:
```diff
@@ -0,0 +1,227 @@
+
+Now we will use the `spin aka crons` command to invoke the Spin app's HTTP endpoint on our desired schedule. The syntax for this request follows standard crontab format.
+
+> `spin aka crons` supports multiple digits in each cron expression field and intervals in every position (e.g., `* */12 * * *` for every 12 hours). Syntax elements like comma-separated lists and ranges are currently supported. Scheduling follows UTC and you may need to [convert your local time to UTC](https://www.worldtimebuddy.com/?pl=1&lid=100&h=100&hf=1) to ensure correct execution during day ...
```
Won't I need to convert my local time to UTC anyway, if I want it to fire at a particular local time?  And then amend it at each DST change?

macolso replied:
Yes, that's a good point. I liked your language suggestion below -- I will make that modification.

itowlson commented on content/wasm-functions/using-cron-jobs.md:
```diff
@@ -0,0 +1,227 @@
+
+Now we will use the `spin aka crons` command to invoke the Spin app's HTTP endpoint on our desired schedule. The syntax for this request follows standard crontab format.
+
+> `spin aka crons` supports multiple digits in each cron expression field and intervals in every position (e.g., `* */12 * * *` for every 12 hours). Syntax elements like comma-separated lists and ranges are currently supported. Scheduling follows UTC and you may need to [convert your local time to UTC](https://www.worldtimebuddy.com/?pl=1&lid=100&h=100&hf=1) to ensure correct execution during day ...
```
"Scheduling follows UTC" did not leave me 100% clear that I understood what I was being told.  I _assume_ this is saying that I must enter times of day in UTC rather than local time (i.e. the system does not convert for me).  Could we say explicitly "if your schedule includes a time of day, that time of day must be in UTC"?  (I mean assuming that _is_what it means!)

rylev replied:
That is indeed what it means.

itowlson commented on content/wasm-functions/using-cron-trigger.md:
```diff
@@ -7,18 +7,20 @@ enable_shortcodes = true
+- [Next Steps](#next-steps)
   
-This tutorial guides you through the process of scheduling HTTP requests in a Spin application running on Fermyon Wasm Functions using cron jobs. The `spin aka cron` command allows applications to receive scheduled HTTP triggers without relying on external services. In this tutorial, you'll learn how to define and manage cron jobs using the `spin aka cron` command.
+This tutorial guides you through the process of scheduling HTTP requests in a Spin application running on Fermyon Wasm Functions using cron jobs. The `spin aka crons` command allows applications to receive scheduled HTTP triggers without relying on external services. In this tutorial, you'll learn how to define and manage cron jobs using the `spin aka crons` command.
```
This page seems to be largely the same content as the previous one but with the "cron trigger" language back in... is this meant to have been deleted in favour of the page above?  The TOC entry still points to this one.

itowlson commented on content/wasm-functions/using-cron-jobs.md:
````diff
@@ -0,0 +1,227 @@
+spin aka deploy
+```
+
+Upon successful deployment, you should receive a domain name that you can use to test the application:
````
Consider shoiwing the deploy output so they can see where to find the domain name.

itowlson reviewed (changes requested):
Thanks for the update @macolso - this is much clearer now - unfortunately I still have a few niggles and typos, and at least one thing which I'm pretty sure is an error.  Sorry for being picky!

macolso commented on content/wasm-functions/using-cron-jobs.md:
```diff
@@ -0,0 +1,227 @@
+        // Capture current timestamp for logging and response
+        const now = new Date().toISOString();
+
+        // Log the trigger event, including the value of 'msg'
```
I decided to simplify the example and remove the functionality that changes the output depending on whether the app received a direct ping or a cron trigger. I agree that the added complexity outweighs the value it's providing.

macolso commented:
@itowlson Excellent feedback—thank you! I've taken another stab at addressing the concerns. I think it flows much better now, and I agree that the simplified example makes it easier to read and keeps the user focused on the task at hand. Let me know what you think

itowlson commented on content/wasm-functions/using-cron-jobs.md:
````diff
@@ -0,0 +1,237 @@
+
+We'll go ahead add a path that will be used our cron job in later steps:
+
+```toml
````
Do we need to add a second trigger?  Is this trying to illustrate a recommended practice?  It would be good to say more about the motivation because right now this feels like a step I follow without knowing why.

macolso replied:
I went back and forth on this... I decided to not add a second trigger to keep the example simple. Perhaps I can write a more realistic example for Spin Hub that we can link as a resource at the end of the tutorial for those interested in that scenario. 


To simplify things, I've removed the specific path and replaced it with wildcard.

itowlson commented on content/wasm-functions/using-cron-jobs.md:
````diff
@@ -0,0 +1,237 @@
+component = "hello-cron-job"
+```
+
+> Note that you can have multiple cron jobs per Spin application as long as the combination of schedule and path_and_query is unique. We will review how to set those values later in the tutorial.  
````
This seems a bit disconnected from "add a trigger."  It feels like you're setting up for a mental model of "multiple cron jobs per app, each cron job handled by a different component, those components on different triggers, cron jobs mapped to components via their routes" but it's not explicit, and the "each trigger points at the same component as the everything route and we route within the component" doesn't draw out that structure.

Perhaps move this down to where you talk about creating cron jobs, and if you want to talk about how to structure a multi-job app then do so there?  And focus on the simple "get something working" path here?  Or do you feel that leads readers to a dead end?

macolso replied:
Agreed - I've moved it down to the section where I talk about the required parameters for the `spin aka crons create` subcommand. I decided to not expand into multi-job to keep the tutorial manageable, but I like the idea of creating a more complex example for Spin Hub or fwf-examples we can link to in this tutorial for those wanting a more realistic example.

itowlson commented on content/wasm-functions/using-cron-jobs.md:
```diff
@@ -0,0 +1,237 @@
+
+* **Schedule** – How often the job runs, in standard cron syntax
+* **Path and query** – The HTTP path (and optional query parameters) to invoke
+* **Name** – A name for your cron job (this will be associated with the Spin application in your current working directory)
```
Consider breaking this up - "it applies to the app in your current directory" is important enough that it shouldn't be a parenthesis under "the name for the job".  E.g.

```suggestion
* **Name** – A name for your cron job

The cron job will always invoke the Spin application in your current working directory.
```

(or perhaps "...will always invoke the Spin application deployed from your current..." which avoids any implication that it will invoke it _in_ your cwd, but is perhaps awkward and requires extra effort for the reader to parse, not sure)

macolso replied:
Good point about highlighting this. I removed it from parenthesis and put it at the beginning of the paragraph for this section.

itowlson commented on content/wasm-functions/using-cron-jobs.md:
```diff
@@ -0,0 +1,237 @@
++----------------+--------------+-------------------------+
+| Name           | Schedule     | Next Run                |
++=========================================================+
+| hello-cron-job |  */5 * * * * | 2025-04-02 04:00:00 UTC |
```
```suggestion
| cron-job-1 |  */5 * * * * | 2025-04-02 04:00:00 UTC |
```

or wait, are you saying this is the app name rather than the job name?

(I am not sure if the command shows _all_ cron jobs or only the ones associated with the current app.)

macolso replied:
d'oh - typo! I fixed this. I also renamed the app to "hello-world" to avoid confusion between app name and cron job name.  

> I am not sure if the command shows all cron jobs or only the ones associated with the current app.)
@rylev any guidance on this? I'm not sure

itowlson replied:
Having thoroughly distinct different names for the app and the cron job is an excellent idea - thank you!

itowlson commented on content/wasm-functions/using-cron-jobs.md:
````diff
@@ -0,0 +1,237 @@
+<!-- @selectiveCpy -->
+
+```console
+spin aka crons delete hello-cron-job
````
```suggestion
spin aka crons delete cron-job-1
```

itowlson commented on content/wasm-functions/using-cron-jobs.md:
````diff
@@ -0,0 +1,237 @@
+<!-- @nocpy -->
+
+```console
+Deleted cron job 'hello-cron-job' with schedule '*/5 * * * *'
````
```suggestion
Deleted cron job 'cron-job-1' with schedule '*/5 * * * *'
```

itowlson reviewed (commented):
Thanks @macolso - I agree that this flows better - unfortunately I did get confused at one point, and spotted a couple of what I think were typos arising from the previous reuse of names.  I may be wrong though...

itowlson commented on content/wasm-functions/using-cron-jobs.md:
````diff
@@ -0,0 +1,226 @@
+<!-- @selectiveCpy -->
+
+```console
+spin aka crons create "*/5 * * * *" "/*" "cron-job-1"
````
```suggestion
spin aka crons create "*/5 * * * *" "/" "cron-job-1"
```

itowlson reviewed (approved):
This looks good to me.  I'm happy to defer the "does it list all jobs or current app's jobs" if you want to get this done and dusted, it's a fairly minor detail and we can always add it later if the spirit moves us.  (Although it may be worth waiting for Ryan anyway!)  Thanks for your patience through the review cycles!

macolso commented:
> This looks good to me. I'm happy to defer the "does it list all jobs or current app's jobs" if you want to get this done and dusted, it's a fairly minor detail and we can always add it later if the spirit moves us. (Although it may be worth waiting for Ryan anyway!) Thanks for your patience through the review cycles!

Thanks for all the polish help! Between writing most of this at conferences or on plane, it definitely needed the help... I will go ahead and merge and we can add the "does it list all jobs or current app's jobs" later. I think it would be helpful to maybe even include an article about the purpose of that config file in our docs since it's a bit of an "insider knowledge" resource ATM.
